src/emulators/avd.js

```diff
@@ -1 +1,334 @@
+ * @returns {Promise}
+ */
+export function isRunning(emu, devices = [], opts = {}) {
+	if (!devices.length || emu.type !== 'avd') {
+		return Promise.resolve(false);
+	}
+	const result = devices.filter(d => { return emuRegExp.test(d.id) && d.emulator && d.emulator.name === emu.name; });
+	return Promise.resolve(result.shift());
+}
+
+/**
+ * Detects if a specific device name is an Android emulator.
+ *
+ * @param {String} deviceId - The id of the emulator.
+ * @param {Object} opts - Various options.
+ * @returns {Promise}
+ */
+export function isEmulator(deviceId, opts = {}) {
+	let port = deviceId.match(emuRegExp);
+	if (!port) {
+		return Promise.resolve(false);
+	}
+
+	return Promise.all([
+		getAvdName(port[1]),
+		this.detect(opts)
+	])
+	.then(([avdName, avdInfo]) => avdInfo && avdInfo.avds.filter(a => { return a.name === avdName; }).shift());
+}
+
+/**
+ * Launches the specified Android emulator.
+ *
+ * @param {Object} emu - An emulator object.
+ * @param {Object} [opts] - Emulator start options.
+ * @param {Boolean} [opts.bypassCache=false] - Bypasses the Android environment detection cache and re-queries the system.
+ * @param {Number} [opts.port=5560] - The TCP port the emulator will use for the console.
+ * @param {String} [opts.sdcard] - A path to the virtual SD card to use with the emulator.
+ * @param {Number} [opts.partitionSize=128] - The emulator's system/data partition size in MBs.
+ * @param {Array|String} [opts.stdio] - The stdio configuration to pass into spawn().
+ * @param {Boolean} [opts.detached] - The detached flag to pass into spawn().
+ * @returns {Promise}
+ */
+export function start(emu, opts = {}) {
+	let port = opts.port || DEFAULTPORT;
+	let emulatorExe;
+
+	return sdk
+		.detect(opts)
+		.then(result => {
+			emulatorExe = result.sdk.executables.emulator;
+
+			function doUntil(p) {
+				return util.findport(p)
+					.then(result => {
+						if (result || p >= PORTLIMIT) {
+							return result;
+						} else {
+							p = p + 2;
+							return doUntil(p);
+						}
+					});
+			}
+
+			return doUntil(port)
+				.then(result => {
+					emu.emit('message', `Found port "${result}" for "${emu.name}"`);
+					return result;
+				});
+		})
+		.then(port => {
+			emu.emit('message', `Using port "${port}"`);
+
+			// default args
+			let args = [
+				'-avd', emu.name,                                // use a specific android virtual device
+				'-port', port,                                   // TCP port that will be used for the console
+				'-no-boot-anim',                                 // disable animation for faster boot
+				'-partition-size', opts.partitionSize || 128     // system/data partition size in MBs
+			];
+
+			let sdcard = opts.sdcard || emu.sdcard;
+			sdcard && args.push('-sdcard', sdcard);              // SD card image (default <system>/sdcard.img
+
+			// add any other args
+			opts.logcat               && args.push('-logcat', opts.logcat);                // enable logcat output with given tags
```

Comment:
Get rid of all this garbage that I wrote and just have a simple:

    if (Array.isArray(opts.extraArgs)) {
      args.push.apply(args, opts.extraArgs);
    }